File: cloud-run-bot/scrapers/getir_scraper.js
/**
 * Getir Scraper (Node.js port)
 */
const BaseProductScraper = require('./base_scraper');

class GetirScraper extends BaseProductScraper {
  get domain() { return 'getir.com'; }

  canHandle(url) {
    return url.toLowerCase().includes('getir.com');
  }

  scrapeImage($, url) {
    // 1. Next.js __NEXT_DATA__
    const pData = this._getProductData($);
    if (pData) {
      const picUrls = pData.picURLs;
      if (Array.isArray(picUrls) && picUrls.length > 0) {
        const img = picUrls[0].toString();
        const resolved = this.resolveImageUrl(img, url);
        if (resolved && !this.isLogoUrl(resolved)) {
          return resolved;
        }
      }

      const sqThumb = pData.squareThumbnailURL;
      if (sqThumb) {
        const resolved = this.resolveImageUrl(sqThumb.toString(), url);
        if (resolved && !this.isLogoUrl(resolved)) {
          return resolved;
        }
      }
    }

    // 2. og:image
    const ogImg = $('meta[property="og:image"]').attr('content') ||
                  $('meta[name="twitter:image"]').attr('content');
    if (ogImg && !this.isLogoUrl(ogImg)) {
      const resolved = this.resolveImageUrl(ogImg, url);
      if (resolved) return resolved;
    }

    return null;
  }

  scrapeTitle($) {
    const pData = this._getProductData($);
    if (pData) {
      if (pData.name && pData.name.trim().length > 0) {
        return pData.name.trim();
      }
      if (pData.shortName && pData.shortName.trim().length > 0) {
        return pData.shortName.trim();
      }
    }

    const ogTitle = $('meta[property="og:title"]').attr('content') || $('title').text();
    if (ogTitle && ogTitle.toLowerCase() !== 'null') {
      let title = ogTitle;
      title = title.replace(/ - Getir/g, '').replace(/ \| Getir/g, '');
      return title.trim();
    }

    return null;
  }

  scrapePrice($) {
    const pData = this._getProductData($);
    if (pData && pData.price != null) {
      const parsed = parseFloat(pData.price);
      if (!isNaN(parsed) && parsed > 0) {
        return parsed;
      }
    }
    return null;
  }

  scrapeDescription($) {
    const pData = this._getProductData($);
    if (pData && pData.shortDescription && pData.shortDescription.trim().length > 0) {
      return pData.shortDescription.trim();
    }

    const ogDesc = $('meta[name="description"], meta[property="og:description"]').first();
    if (ogDesc.length) {
      const content = ogDesc.attr('content')?.trim();
      if (content && content.toLowerCase() !== 'null') return content;
    }
    return null;
  }

  scrapeBreadcrumbs($) {
    const pData = this._getProductData($);
    if (!pData) return [];

    const breadcrumbs = [];
    const cats = this._getCategoriesList($);
    if (Array.isArray(cats) && cats.length > 0) {
      const productCatIds = pData.categoryIds;
      const productSubcatIds = pData.subCategoryIds;
      if (Array.isArray(productCatIds)) {
        for (const cat of cats) {
          if (cat && typeof cat === 'object' && productCatIds.includes(cat.id)) {
            if (cat.name) breadcrumbs.push(cat.name);
            const subcats = cat.subCategories;
            if (Array.isArray(subcats) && Array.isArray(productSubcatIds)) {
              for (const sub of subcats) {
                if (sub && typeof sub === 'object' && productSubcatIds.includes(sub.id)) {
                  if (sub.name) breadcrumbs.push(sub.name);
                }
              }
            }
          }
        }
      }
    }

    return breadcrumbs;
  }

  _getProductData($) {
    const scripts = $('script');
    for (let i = 0; i < scripts.length; i++) {
      const id = $(scripts[i]).attr('id');
      const text = $(scripts[i]).text() || '';
      if (id === '__NEXT_DATA__' || text.includes('"productDetail"')) {
        try {
          const data = JSON.parse(text);
          const pData = this._findProductDetailData(data);
          if (pData) return pData;
        } catch (_) {}
      }
    }
    return null;
  }

  _findProductDetailData(json) {
    if (json && typeof json === 'object' && !Array.isArray(json)) {
      if (json.productDetail && typeof json.productDetail === 'object') {
        const pd = json.productDetail;
        if (pd.data && typeof pd.data === 'object' && pd.data.name && pd.data.price != null) {
          return pd.data;
        }
      }
      for (const value of Object.values(json)) {
        const res = this._findProductDetailData(value);
        if (res) return res;
      }
    } else if (Array.isArray(json)) {
      for (const item of json) {
        const res = this._findProductDetailData(item);
        if (res) return res;
      }
    }
    return null;
  }

  _getCategoriesList($) {
    const scripts = $('script');
    for (let i = 0; i < scripts.length; i++) {
      const id = $(scripts[i]).attr('id');
      const text = $(scripts[i]).text() || '';
      if (id === '__NEXT_DATA__' || text.includes('"getirListing"')) {
        try {
          const data = JSON.parse(text);
          return this._findCategoriesList(data);
        } catch (_) {}
      }
    }
    return null;
  }

  _findCategoriesList(json) {
    if (json && typeof json === 'object' && !Array.isArray(json)) {
      if (json.getirListing && typeof json.getirListing === 'object') {
        const gl = json.getirListing;
        if (gl.categories && typeof gl.categories === 'object' && Array.isArray(gl.categories.data)) {
          return gl.categories.data;
        }
      }
      for (const value of Object.values(json)) {
        const res = this._findCategoriesList(value);
        if (res) return res;
      }
    } else if (Array.isArray(json)) {
      for (const item of json) {
        const res = this._findCategoriesList(item);
        if (res) return res;
      }
    }
    return null;
  }
}

module.exports = GetirScraper;

File: lib/services/scrapers/getir_scraper.dart
import 'dart:convert';
import 'package:html/dom.dart' as dom;
import 'base_scraper.dart';

class GetirScraper extends BaseProductScraper {
  @override
  String get domain => 'getir.com';

  @override
  bool canHandle(String url) {
    return url.toLowerCase().contains('getir.com');
  }

  @override
  String? scrape({
    required dom.Document document,
    required String url,
    required bool Function(String urlString) isLogoUrl,
    required String? Function(String? imageUrl, String pageUrl) resolveImageUrl,
    required void Function(String message) log,
  }) {
    // 1. Next.js JSON-LD/Next data (Öncelikli)
    final pData = _getProductData(document);
    if (pData != null) {
      // Önce picURLs listesindeki ilk resmi almayı dene
      final picUrls = pData['picURLs'];
      if (picUrls is List && picUrls.isNotEmpty) {
        final img = picUrls.first.toString();
        final resolved = resolveImageUrl(img, url);
        if (resolved != null && !isLogoUrl(resolved)) {
          log('   - Getir görseli JSON picURLs ile bulundu: $resolved');
          return resolved;
        }
      }
      
      // Alternatif olarak squareThumbnailURL
      final sqThumb = pData['squareThumbnailURL'];
      if (sqThumb != null && sqThumb.toString().isNotEmpty) {
        final resolved = resolveImageUrl(sqThumb.toString(), url);
        if (resolved != null && !isLogoUrl(resolved)) {
          log('   - Getir görseli JSON squareThumbnailURL ile bulundu: $resolved');
          return resolved;
        }
      }
    }

    // 2. og:image meta tag (Fallback)
    final ogImage = document.querySelector('meta[property="og:image"]')?.attributes['content'] ??
                    document.querySelector('meta[name="twitter:image"]')?.attributes['content'];
    if (ogImage != null && ogImage.isNotEmpty) {
      final resolved = resolveImageUrl(ogImage, url);
      if (resolved != null && !isLogoUrl(resolved)) {
        log('   - Getir görseli og:image ile bulundu: $resolved');
        return resolved;
      }
    }

    return null;
  }

  @override
  String? scrapeTitle(dom.Document document) {
    final pData = _getProductData(document);
    if (pData != null) {
      final name = pData['name'];
      if (name is String && name.isNotEmpty) {
        return name.trim();
      }
      final shortName = pData['shortName'];
      if (shortName is String && shortName.isNotEmpty) {
        return shortName.trim();
      }
    }

    final ogTitle = document.querySelector('meta[property="og:title"]')?.attributes['content'] ??
                    document.querySelector('title')?.text;
    if (ogTitle != null && ogTitle.isNotEmpty && ogTitle.toLowerCase() != 'null') {
      String title = ogTitle;
      if (title.contains(' - Getir')) {
        title = title.replaceAll(' - Getir', '');
      }
      if (title.contains(' | Getir')) {
        title = title.replaceAll(' | Getir', '');
      }
      return title.trim();
    }

    return null;
  }

  @override
  Future<double?> scrapePrice(dom.Document document) async {
    final pData = _getProductData(document);
    if (pData != null) {
      final priceVal = pData['price'];
      if (priceVal != null) {
        final parsed = double.tryParse(priceVal.toString());
        if (parsed != null && parsed > 0) {
          return parsed;
        }
      }
    }
    return null;
  }

  @override
  String? scrapeDescription(dom.Document document) {
    final pData = _getProductData(document);
    if (pData != null) {
      final shortDesc = pData['shortDescription'];
      if (shortDesc is String && shortDesc.isNotEmpty) {
        return shortDesc.trim();
      }
    }

    final ogDesc = document.querySelector('meta[name="description"]')?.attributes['content'] ??
                   document.querySelector('meta[property="og:description"]')?.attributes['content'];
    if (ogDesc != null && ogDesc.isNotEmpty && ogDesc.toLowerCase() != 'null') {
      return ogDesc.trim();
    }

    return null;
  }

  @override
  List<String> scrapeBreadcrumbs(dom.Document document) {
    final pData = _getProductData(document);
    if (pData == null) return [];
    
    final List<String> breadcrumbs = [];
    
    // GetirListing altındaki kategorileri bulup product categoryIds/subCategoryIds ile eşleştirme
    final cats = _getCategoriesList(document);
    if (cats != null && cats.isNotEmpty) {
      final productCatIds = pData['categoryIds'];
      final productSubcatIds = pData['subCategoryIds'];
      if (productCatIds is List) {
        for (final cat in cats) {
          if (cat is Map && productCatIds.contains(cat['id'])) {
            final catName = cat['name'];
            if (catName is String && catName.isNotEmpty) {
              breadcrumbs.add(catName);
            }
            final subcats = cat['subCategories'];
            if (subcats is List && productSubcatIds is List) {
              for (final sub in subcats) {
                if (sub is Map && productSubcatIds.contains(sub['id'])) {
                  final subName = sub['name'];
                  if (subName is String && subName.isNotEmpty) {
                    breadcrumbs.add(subName);
                  }
                }
              }
            }
          }
        }
      }
    }
    
    return breadcrumbs;
  }

  Map<String, dynamic>? _getProductData(dom.Document document) {
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final id = script.attributes['id'];
      if (id == '__NEXT_DATA__' || script.text.contains('"productDetail"')) {
        try {
          final data = jsonDecode(script.text);
          final pData = _findProductDetailData(data);
          if (pData != null) return pData;
        } catch (_) {}
      }
    }
    return null;
  }

  Map<String, dynamic>? _findProductDetailData(dynamic json) {
    if (json is Map) {
      if (json.containsKey('productDetail') && json['productDetail'] is Map) {
        final pd = json['productDetail'];
        if (pd is Map && pd.containsKey('data') && pd['data'] is Map) {
          final pData = pd['data'];
          if (pData.containsKey('name') && pData.containsKey('price')) {
            return Map<String, dynamic>.from(pData);
          }
        }
      }
      for (final value in json.values) {
        final res = _findProductDetailData(value);
        if (res != null) return res;
      }
    } else if (json is List) {
      for (final item in json) {
        final res = _findProductDetailData(item);
        if (res != null) return res;
      }
    }
    return null;
  }

  List<dynamic>? _getCategoriesList(dom.Document document) {
    final scripts = document.querySelectorAll('script');
    for (final script in scripts) {
      final id = script.attributes['id'];
      if (id == '__NEXT_DATA__' || script.text.contains('"getirListing"')) {
        try {
          final data = jsonDecode(script.text);
          return _findCategoriesList(data);
        } catch (_) {}
      }
    }
    return null;
  }

  List<dynamic>? _findCategoriesList(dynamic json) {
    if (json is Map) {
      if (json.containsKey('getirListing') && json['getirListing'] is Map) {
        final gl = json['getirListing'];
        if (gl is Map && gl.containsKey('categories') && gl['categories'] is Map) {
          final cats = gl['categories'];
          if (cats is Map && cats.containsKey('data') && cats['data'] is List) {
            return cats['data'] as List;
          }
        }
      }
      for (final value in json.values) {
        final res = _findCategoriesList(value);
        if (res != null) return res;
      }
    } else if (json is List) {
      for (final item in json) {
        final res = _findCategoriesList(item);
        if (res != null) return res;
      }
    }
    return null;
  }
}
